Disallow duplicate thoughts

pushkov-fedor opened the pull request:
Fixes #764

raineorshine commented:
@shresthabijay I'd like your opinion since your familiar with this part of the code. Is there a better way to change the opacity of the Editable when a duplicate value is detected? I'm hesitant to use direct DOM manipulation. Is there a more idiomatic way to do this in React? Thanks!

shresthabijay commented:
> @shresthabijay I'd like your opinion since your familiar with this part of the code. Is there a better way to change the opacity of the Editable when a duplicate value is detected? I'm hesitant to use direct DOM manipulation. Is there a more idiomatic way to do this in React? Thanks!

I don't think there is any better solution to this. The problem is that we are restricted to not re-render `Editable`. I really haven't looked into why we cannot re-render `Editable` but I guess it has something to do with focus selection. I really think we should solve that issue instead of trying to find workaround to direct DOM manipulation.

raineorshine commented:
@shresthabijay You're right, but that seems unavoidable. We can't re-render an Editable while the user is editing it, otherwise it would mess up their focus.

If only a style or className is changed, does that still re-render an element?

pushkov-fedor commented:
> @shresthabijay You're right, but that seems unavoidable. We can't re-render an Editable while the user is editing it, otherwise it would mess up their focus.
> 
> If only a style or className is changed, does that still re-render an element?

I think in our case we need to change appearance of element, which could be done by direct dom manipulation or via render function. If we use idiomatic "React" way, all changing in appearance happen after execution of render function, which one we can't use. Correct me please, if i'm wrong somewhere,

raineorshine commented:
Yes, Bijay helped clarify that we *should* be able to use the render function without interfering with editing, but due to the current architecture it forces the DOM element to get re-rendered for some reason. That will need to get addressed in a separate issue.

For now, then, we can use direct DOM manipulation. The whole file deserves some refactoring in the future anyway.

pushkov-fedor commented:
I think that we need to handle this situation in `newThought` action-creator. As i understand, correct result after hitting `Enter` is:
- ba
- bac
- <empty>

I suggest to use `cursor` and `editingValue`. Check if head value of cursor starts with editing value and length of head value is more than length of editingValue. After that check if we have duplicates in tail context of cursor. If these three checks return true then instead of dispatching `splitThought` we dispatch `newThought`.
I can add small function in `newThought` action-creator and invoke it at the end of  
`const split = !preventSplit && cursor && isFocusOnEditable && !showContexts && offset > 0 && offset < headValue(cursor).length`

What do you think?

raineorshine commented:
First of all, thank you for the clear explanation and proposal. That's exactly what I'm looking for.

I agree that this should happen in `newThought` where the `splitThought` condition is applied.

> I think that we need to handle this situation in `newThought` action-creator. As i understand, correct result after hitting `Enter` is:
> 
> * ba
> * bac
> * 

Yes, that's correct.

> Check if head value of cursor starts with editing value and length of head value is more than length of editingValue. 

It is incorrect to assume that the edit must come at the end of the thought. e.g. This would not the case of deleting the `c` from `bca`.

> After that check if we have duplicates in tail context of cursor. 

We shouldn't have to explicitly check for duplicates. It should only be a matter of the browser selection position.

> What do you think?

I think there is an easier solution. According to `state.thoughts`, the cursor thought has not been edited yet. According to the DOM and browser selection it has. The issue occurs because both are being used as a source of truth. See if you can find a way to change the condition for `splitChain` by only looking at the `editingValue`.

> Are there any other actions that depend on the caret position that may be affected as well?

This is important, too. You may have to search for the use of `offset`.

pushkov-fedor commented:
As i understand, hitting 'Enter' on duplicate Thought should reset its value to the last not duplicate and trigger `newThought` action. In this way, the simplest approach to do it - we can add 'isDuplicate' attribute to editable div and retrieve its value in `newThought` action-creator. After checking this value we can decide either invoke `splitThought` or invoke `newThought`. But it's not 'React' way too.
I'm not sure that understand your way to determining what should be called `splitThought` or `newThought` just from browser selection and offset, because they don't tell us about duplicates

raineorshine commented:
> As i understand, hitting 'Enter' on duplicate Thought should reset its value to the last not duplicate and trigger `newThought` action. 

Yes, although it should be noted that the "reset" is DOM-only; the state value was never actually changed.

> In this way, the simplest approach to do it - we can add 'isDuplicate' attribute to editable div and retrieve its value in `newThought` action-creator. After checking this value we can decide either invoke `splitThought` or invoke `newThought`. But it's not 'React' way too.
> I'm not sure that understand your way to determining what should be called `splitThought` or `newThought` just from browser selection and offset, because they don't tell us about duplicates

I still don't think it's the duplicate that matters here. That is a contingency. The general situation is that when the `editingValue` differs from the `cursor` value, then we should use the `editingValue` to detect if the thought should be split or not.

How about we each implement our preferred solution, and then compare the results?

pushkov-fedor commented:
> How about we each implement our preferred solution, and then compare the results?

I think it would be fun, let's try it :)

raineorshine commented:
Okay, my solution is in ee1f2528 ;)

pushkov-fedor commented:
Looks like your won :)

raineorshine commented:
Yours is quite clean! But yes, I think the generality and lack of additional DOM manipulation is preferred.

raineorshine commented:
Just realized that [merge-on-delete](https://github.com/cybersemics/em/blob/dev/src/reducers/deleteEmptyThought.ts#L60-L108) (inverse of split) can be used to create a duplicate:

## Steps to Reproduce

- ab
- a
- b

With the caret at the beginning (offset: 0) of `b`, hit Backspace.

## Current Behavior

- ab
- ab

## Expected Behavior

Maybe prevent the action completely and show an appropriate alert? What do you think?

pushkov-fedor commented:
> Maybe prevent the action completely and show an appropriate alert? What do you think?

Yes, i think we can handle this case in deleteEmptyThoughtOrOutdent shortcut and add another check.

raineorshine commented:
Yes, let's do that.

raineorshine commented:
There is a small regression:

## Steps to Reproduce

- a
- [empty]

Try to delete the empty thought.

## Current Behavior

`Duplicate thoughts are not allowed within the same context`

## Expected Behavior

Empty thought should be deleted.

pushkov-fedor commented:
> There is a small regression:
> 
> ## Steps to Reproduce
> * a
> * [empty]
> 
> Try to delete the empty thought.
> 
> ## Current Behavior
> `Duplicate thoughts are not allowed within the same context`
> 
> ## Expected Behavior
> Empty thought should be deleted.

Fixed, thanks!

raineorshine commented:
- When `Enter` is hit on a duplicate thought, the "Duplicate thoughts are not allowed" alert is not closed.

raineorshine commented:
Superscript regression:

## Steps to Reproduce

- a
  - m
- b
  - mm
  - m| [cursor]

Edit `/b/m` →  `b/mm`

## Current Behavior

Superscript position is not updated.

## Expected Behavior

Superscript should be positioned correctly at the end of the thought text.

pushkov-fedor commented:
@raineorshine, hi! The problem was in inner html of subthought-text div. I set editingValue on focus event in order to fix this case:
- a
  - m
- b
  - mm
  - m| [cursor]

`Edit /b/m → b/mm, hit Enter, then set cursor again on /b/m`

I use editingValue for subthought-text, which in this case is `mm` instead of `m`

pushkov-fedor commented on src/components/Editable.tsx:
```diff
@@ -346,6 +374,10 @@ const Editable = ({ disabled, isEditing, thoughtsRanked, contextChain, cursorOff
+    // dispatch in timeout to avoid "Error: Reducers may not dispatch actions.""
+    setTimeout(() => {
+      dispatch({ type: 'alert', value: null, alertType: 'duplicateThoughts' })
+    }, 4000)
```
It was necessary in order to dispatch null alert only after other duplication alert had been dispatched after 2 seconds delay. I removed delay at all and a bit reorganized alert dispatching logic.

pushkov-fedor commented on src/components/Editable.tsx:
```diff
@@ -269,6 +273,11 @@ const Editable = ({ disabled, isEditing, thoughtsRanked, contextChain, cursorOff
+      if (contentRef.current) {
+        contentRef.current.style.opacity = '1.0'
+      }
+      duplicateThoughtsAlertTimeout.current && window.clearTimeout(duplicateThoughtsAlertTimeout.current)
```
We need to store timeout id somewhere (not in state) in order to clear timeout if user has changed duplicate thought value while alert about duplication is waiting to be dispatched in timeout

raineorshine replied:
Thanks! But why in a ref? Why not just use closure since they are only ever editing one thought at a time?

pushkov-fedor replied:
We use this ref in different functions across Editable and its value could change (if during editing we detect 2+ duplication cases). I can write some utility function for handling duplication alert, define there timeoutId variable and return another function which invokes setTimeout and sets timeoutId, but it needs using let instead of const. Maybe i don't understand something about your idea? Correct me please, if i'm wrong somewhere.

raineorshine replied:
ref is used specifically to tie functionality to the mount/unmount cycle. I don't yet see how the duplicate functionality requires a ref. You can only edit one item at a time, so I don't see how 2+ duplicate cases are relevant.

A ref is more complicated than a simple `let` with closure, so I'm asking for you to provide justification for using this more complex structure. `const` is misleading because the ref has internal state that gets mutated, so it is no more functional than `let`. They both use state, albeit different kinds. Whenever cancelling a timeout is needed, state is involved.

What is stopping you from placing a variable with `let` at a high enough scope so that is accessible where it is needed? Is there a case that requires ref so that is bound to the mount/unmount cycle that I am not aware of? Again, I might be missing something, so let me know! Thanks!

pushkov-fedor replied:
Yes, you are right. I didn't want to use let, but you made a good point, that ref and let are actually the same in term of mutating. I replaced ref with closure and was working on fixing this:
- a
  - ba
  - bac|

After hitting `Backspace` with the caret at the end of **bac** and then hitting `Enter`, alert is dispatched and doesn't hide after a while. I realized that hitting `Enter` triggers re-render component before onBlur and return function of useEffect are fired. After that closure is re-created and we lose timeout.

raineorshine commented:
Better, thanks!

I found this:

## Steps to Reproduce

- a
  - b
  - b|

After typing the second `b`, hit Enter and wait 2 seconds.

## Current Behavior

Duplicate thought alert incorrectly displays after 2 seconds.

## Expected Behavior

Duplicate thought alert should not be displayed since the cursor moved and there are no duplicates.

raineorshine commented:
- https://github.com/cybersemics/em/pull/786#discussion_r463873914
- merge conflicts

raineorshine commented on src/components/Editable.tsx:
```diff
@@ -502,6 +502,10 @@ const Editable = ({ disabled, isEditing, thoughtsRanked, contextChain, cursorOff
    */
   const onKeyDown = (e: React.KeyboardEvent) => {
     if (e.key in MODIFIER_KEYS) return
+    if (e.key === 'Enter' && duplicateThoughtsAlertTimeout.current) {
+      window.clearTimeout(duplicateThoughtsAlertTimeout.current)
+      duplicateThoughtsAlertTimeout.current = undefined
+    }
```
This doesn't seem like the right place for this. Can you clear the alert in blur or unmount?

pushkov-fedor replied:
Yes, you're right

raineorshine commented:
It appears that there is an indiscriminate `alert(null)` somewhere that is clearing the undo alert. Archive a thought and you will see the alert appear and then disappear immediately.

raineorshine commented:
There is a duplicate produced in the following scenario: 

Hit <kbd>Enter</kbd> to split `ab`, when `a` or `b` are siblings.

- a
- a|b

pushkov-fedor commented:
> There is a duplicate produced in the following scenario:
> 
> Hit Enter to split `ab`, when `a` or `b` are siblings.
> 
> * a
> * a|b

This case is kind of similar with previous splitting case:
> There is a strange edge case that we need to address:
> - a
>   - ba
>   - bac|
>
> Hit Backspace with the caret at the end of bac and then hit Enter. This will create a duplicate ba in the same context. The cause is that Enter incorrectly triggers splitThought because focusOffset is 2 and value.length is 3 since it has not been modified.

But this time i think the way to fix our problem is to compare splitted thoughts, based on `editingValue`, with other siblings. But this solution overlaps your, based on `offset` and `editingValue.length`

raineorshine commented:
> But this time i think the way to fix our problem is to compare splitted thoughts, based on `editingValue`, with other siblings. But this solution overlaps your, based on `offset` and `editingValue.length`

Yeah, you're right. In this case there is no way to know if it would create a duplicate without explicitly checking for duplicates. You will need to compare the left and right parts of the split `editingValue` with siblings.

raineorshine commented:
Here are all of the issues we encountered along the way. Only one is not fixed. We're almost there!

- [x] [split thought 1](https://github.com/cybersemics/em/pull/786#pullrequestreview-450082400)
- [x] [split thought 2](https://github.com/cybersemics/em/pull/786#issuecomment-670222851)
- [x] [merge on delete](https://github.com/cybersemics/em/pull/786#issuecomment-662750292)
- [x] [delete empty thought](https://github.com/cybersemics/em/pull/786#issuecomment-663562044)
- [x] [superscript position](https://github.com/cybersemics/em/pull/786#issuecomment-664493836)
- [x] [alert timing](https://github.com/cybersemics/em/pull/786#pullrequestreview-458818614)
- [ ] [alert should be dismissed after re-render](https://github.com/cybersemics/em/pull/786#issuecomment-667415193)

raineorshine commented:
Found one small problem:

## Steps to Reproduce

- a
  - b
  - b|

1. Wait for the alert to render after typing the duplicate
2. Hit Enter

## Current Behavior

The alert is not dismissed

raineorshine commented on src/action-creators/newThought.ts:
```diff
@@ -37,11 +36,15 @@ const newThought = ({ offset, preventSplit, value = '' }: { offset: number, prev
 
   const showContexts = cursor && isContextViewActive(state, contextOf(pathToContext(cursor)))
 
+  const context = cursor && (showContexts && cursor.length > 2 ? pathToContext(contextOf(contextOf(cursor)))
+    : !showContexts && cursor.length > 1 ? pathToContext(contextOf(cursor))
+    : [ROOT_TOKEN])
+  const siblings = context && getThoughts(state, context)
+  const isDuplicateOnSplit = siblings && editingValue && siblings.some(sibling => sibling.value === editingValue.substring(0, offset) || sibling.value === editingValue.substring(offset))
```
Let's make `isDuplicateOnSplit` a function (in the same scope, not in a separate file). That will provide better encapsulation of its inputs and a helpful JSDOC.

pushkov-fedor commented:
After rebase on https://github.com/cybersemics/em/pull/823, we have encountered another problem. Try again this case: https://github.com/cybersemics/em/pull/786#pullrequestreview-450082400. Duplicate thought isn't saved in database and `newThought` triggers instead of `splitThought`, but content of Editable doesn't reset after hitting enter and stays the same (as if thought would be saved)

raineorshine commented:
> After rebase on #823, we have encountered another problem. Try again this case: [#786 (review)](https://github.com/cybersemics/em/pull/786#pullrequestreview-450082400). Duplicate thought isn't saved in database and `newThought` triggers instead of `splitThought`, but content of Editable doesn't reset after hitting enter and stays the same (as if thought would be saved)

Okay, good to know. Here are the questions I would ask of you before I did any investigation:

- Have you identified the cause of the problem?
- What are possible solutions?
- What solution do you recommend and why?

Merge regressions are a normal part of development on a team. Since @shresthabijay is the author of #823, I suggest reaching out to him in case he has any insight.

Does that make sense as far as next steps, who to talk to, and when to get me involved?

pushkov-fedor commented:
I didn't dive in this PR and its changes. If we replace custom `ContentEditable` with `react-contenteditable`, everything is okay. I can figure out by myself what happens there, but i would like to get @shresthabijay opinion what could be the reason of such behavior. I think it fixes in a few lines of code.

@shresthabijay Could you help me with this please?
# Steps to Reproduce
Create this tree:
- ba
- bac

Hit `Backspace` with the caret at the end of `bac` and then hit `Enter`. We detect duplicate in `onChangeHandler` and prevent calling `thoughtChangeHandler`
# Current Behavior
- ba
- ba (duplicate isn't saved in db, it is still `bac`, but renders as `ba`
- 
# Expected Behavior
- ba
- bac
-

pushkov-fedor commented:
> Does that make sense as far as next steps, who to talk to, and when to get me involved?

Yes everything is clear!

raineorshine commented:
> I didn't dive in this PR and its changes. If we replace custom `ContentEditable` with `react-contenteditable`, everything is okay. I can figure out by myself what happens there, but i would like to get @shresthabijay opinion what could be the reason of such behavior. I think it fixes in a few lines of code.

Yes, if it is indeed a regression from the custom `ContentEditable`, then @shresthabijay is responsible for that.

shresthabijay commented:
> I didn't dive in this PR and its changes. If we replace custom `ContentEditable` with `react-contenteditable`, everything is okay. I can figure out by myself what happens there, but i would like to get @shresthabijay opinion what could be the reason of such behavior. I think it fixes in a few lines of code.
> 
> @shresthabijay Could you help me with this please?
> 
> # Steps to Reproduce
> Create this tree:
> 
> * ba
> * bac
> 
> Hit `Backspace` with the caret at the end of `bac` and then hit `Enter`. We detect duplicate in `onChangeHandler` and prevent calling `thoughtChangeHandler`
> 
> # Current Behavior
> * ba
> * ba (duplicate isn't saved in db, it is still `bac`, but renders as `ba`
> 
> # Expected Behavior
> * ba
> * bac

@pushkov-fedor `ContentEditable` sets `innerHTML` to latest value only on `blur` and when `Editable` component re-renders with new `thoughtsRanked`. The intention is to prevent caret position to reset on each edit. So for this case you may need to set `innerHTML` to the `oldValue` on duplicate using `contentRef`. Please be sure to rebase to the `dev` before making the changes. It has few new changes to `ContentEditable`.

pushkov-fedor commented:
> @pushkov-fedor `ContentEditable` sets `innerHTML` to latest value only on `blur` and when `Editable` component re-renders with new `thoughtsRanked`. The intention is to prevent caret position to reset on each edit. So for this case you may need to set `innerHTML` to the `oldValue` on duplicate using `contentRef`. Please be sure to rebase to the `dev` before making the changes. It has few new changes to `ContentEditable`.

Thank you for explanation!

pushkov-fedor commented:
@raineorshine Raine, all issues are fixed.

raineorshine commented:
- [x] [split thought 1](https://github.com/cybersemics/em/pull/786#pullrequestreview-450082400)
- [ ] [split thought 2](https://github.com/cybersemics/em/pull/786#issuecomment-670222851)

## Steps to Reproduce

Hit <kbd>Enter</kbd> to split `ab`, when `a` or `b` are siblings.

- a
- a|b

## Current Behavior

- a
- ab
- [empty]

## Expected Behavior

Duplicate alert should be shown, the cursor and caret should not move, and the new thought should not be created. 

This would be more in alignment with user's expectations than the current implementation, which does not provide any message explaining what happened. When there are many children, the user may not even be aware that a duplicate thought would be created if they split the cursor thought. This same approach should probably be taken with any operation other than editing or navigating on a thought in an invalid duplicate state (e.g. indent, outdent, subcategorize, etc).

- [x] [merge on delete](https://github.com/cybersemics/em/pull/786#issuecomment-662750292)
- [x] [delete empty thought](https://github.com/cybersemics/em/pull/786#issuecomment-663562044)
- [x] [superscript position](https://github.com/cybersemics/em/pull/786#issuecomment-664493836)
- [ ] [alert timing](https://github.com/cybersemics/em/pull/786#pullrequestreview-458818614)

After blur, the text resets but the superscript does not. This might involve @shresthabijay again.

![em - 786 blur mov](https://user-images.githubusercontent.com/750276/90824942-6eae4100-e2f5-11ea-8fd5-a4e419f71b40.gif)

- [x] [alert should be dismissed after re-render](https://github.com/cybersemics/em/pull/786#issuecomment-667415193)

pushkov-fedor commented:
> Yes, I think you're right. It will be better for the user to get the alert in this case, too. However, we should probably prevent the action completely. There's no point creating an empty thought when the user's intention was to split the thought. Better to prevent the action, instead of effectively preventing the action AND adding an unwanted action. Does that make sense?

Yes, i have fixed it. Could you check it please?

Also, i have fixed another issue with render. When we add empty thought, edit it and then blur, `innerHTML` of this thought resets to initial empty.

raineorshine commented:
- [ ] [split thought 1](https://github.com/cybersemics/em/pull/786#pullrequestreview-450082400)

The behavior is great, and the alert shows up correctly, but the alert doesn't disappear when it is no longer a duplicate. The alert should always be cleared as soon as the cursor moves or it is edited to no longer be a duplicate.

The other cases all look good:

- [x] [split thought 2](https://github.com/cybersemics/em/pull/786#issuecomment-670222851)
- [x] [merge on delete](https://github.com/cybersemics/em/pull/786#issuecomment-662750292)
- [x] [delete empty thought](https://github.com/cybersemics/em/pull/786#issuecomment-663562044)
- [x] [superscript position](https://github.com/cybersemics/em/pull/786#issuecomment-664493836)
- [x] [alert timing](https://github.com/cybersemics/em/pull/786#pullrequestreview-458818614)
- [x] [alert should be dismissed after re-render](https://github.com/cybersemics/em/pull/786#issuecomment-667415193) 

> Also, i have fixed another issue with render. When we add empty thought, edit it and then blur, `innerHTML` of this thought resets to initial empty.

Good catch. You'll notice it actually still occurs when you blur before `thoughtChangedThrottled` is flushed, i.e. type and then blur within 1 second. This usually means there was a mixup between `cursor` (which changes immediately) and `cursorBeforeEdit` (which changes after the flush).

pushkov-fedor commented:
> Thanks! The behavior is great, and the alert shows up correctly, but the alert doesn't disappear when it is no longer a duplicate. The alert should always be cleared as soon as the cursor moves or it is edited to no longer be a duplicate.

For [split thought 1](https://github.com/cybersemics/em/pull/786#pullrequestreview-450082400) it works this way for me:
![ezgif-6-8837b7421440](https://user-images.githubusercontent.com/32064251/91168110-43fe2700-e6ff-11ea-863f-a9e9686cfbc7.gif)

Alert disappears if thought isn't duplicate or cursor moves. Is it not working in the same way for you? I think i just don't see the case when alert doesn't clear.

pushkov-fedor commented:
> Good catch. You'll notice it actually still occurs when you blur before thoughtChangedThrottled is flushed, i.e. type and then blur within 1 second. This usually means there was a mixup between cursor (which changes immediately) and cursorBeforeEdit (which changes after the flush).

Yes, thanks for notice. I will take a look.

raineorshine commented:
> Alert disappears if thought isn't duplicate or cursor moves. Is it not working in the same way for you? I think i just don't see the case when alert doesn't clear.

Here's a clearer set of steps to reproduce:

1. Create `a`
2. Create `b`
3. Create `b` (duplicate)
4. Hit Enter (duplicate alert will show)
5. Type `c` (so the thought is now `bc`)

You will need to execute steps 3, 4, and 5 within the thought flush delay.

pushkov-fedor commented:
> Here's a clearer set of steps to reproduce:
> 
> 1. Create `a`
> 2. Create `b`
> 3. Create `b` (duplicate)
> 4. Hit Enter (duplicate alert will show)
> 5. Type `c` (so the thought is now `bc`)
> 
> You will need to execute steps 3, 4, and 5 within the thought flush delay.

Fixed! Thank you

pushkov-fedor commented:
There is one case i would like to discuss. Maybe it would be better for user experience if we, instead of creating a new thought, show alert and wait until user edits duplicate thought or moves cursor. What do you think?

> Better, thanks!
> 
> I found this:
> 
> ## Steps to Reproduce
> * a
>   
>   * b
>   * b|
> 
> After typing the second `b`, hit Enter and wait 2 seconds.
> 
> ## Current Behavior
> Duplicate thought alert incorrectly displays after 2 seconds.
> 
> ## Expected Behavior
> Duplicate thought alert should not be displayed since the cursor moved and there are no duplicates.

pushkov-fedor commented:
Couldn't figure out what is wrong with falling test. I have tested it locally, and this test fails/passes from time to time. Last force push doesn't differ from last falling commit, but test is passed.

raineorshine commented:
> There is one case i would like to discuss. Maybe it would be better for user experience if we, instead of creating a new thought, show alert and wait until user edits duplicate thought or moves cursor. What do you think?

I agree. Thank you for thinking about this from the user's perspective, I really appreciate it!

> Couldn't figure out what is wrong with falling test. I have tested it locally, and this test fails/passes from time to time. Last force push doesn't differ from last falling commit, but test is passed.
> 
> UPDATE: I have pushed fetched dev to my repo, and the same tests fail
> [pushkov-fedor/em/runs/1031069346](https://github.com/pushkov-fedor/em/runs/1031069346)

Looks like these tests were failing in `dev`, so nothing you did. It was from my code. I didn't notice it since they involved a subtle concurrency issue and were passing most of the time. I can't verify that I fixed it completely, but I did make some improvements in 616318762, afacf3e09, and 434882035. The tests are passing now in `dev`. If we run into more intermittent failures, we can disable those tests until we have found a proper solution.

pushkov-fedor commented:
> There is one case i would like to discuss. Maybe it would be better for user experience if we, instead of creating a new thought, show alert and wait until user edits duplicate thought or moves cursor. What do you think?

Fixed

pushkov-fedor commented:
Yes, you're right. The problem was in these 2 lines:
```
dispatch({ type: 'alert', value: 'Duplicate thoughts are not allowed within the same context.', alertType: 'duplicateThoughts' })
setTimeout(() => dispatch({ type: 'alert', value: null, alertType: 'duplicateThoughts' }), 2000)
```
Maybe my solution is overcomplicated, and we just can prevent execution of these lines, if duplicate alert has been already dispatched. In this case after 2 second alert will be dismissed. But If user hits `Enter` again after 1.9s of previous hitting, alert will be dismissed in .1s

raineorshine commented:
> Yes, you're right. The problem was in these 2 lines:
> 
> ```
> dispatch({ type: 'alert', value: 'Duplicate thoughts are not allowed within the same context.', alertType: 'duplicateThoughts' })
> setTimeout(() => dispatch({ type: 'alert', value: null, alertType: 'duplicateThoughts' }), 2000)
> ```
> 
> Maybe my solution is overcomplicated, and we just can prevent execution of these lines, if duplicate alert has been already dispatched. In this case after 2 second alert will be dismissed. But If user hits `Enter` again after 1.9s of previous hitting, alert will be dismissed in .1s

The `alert` action can be dispatched again without consequence. It's the `setTimeout` that's the problem. The old `setTimeout` needs to be cleared before it is set again.

pushkov-fedor commented:
> > Yes, you're right. The problem was in these 2 lines:
> > ```
> > dispatch({ type: 'alert', value: 'Duplicate thoughts are not allowed within the same context.', alertType: 'duplicateThoughts' })
> > setTimeout(() => dispatch({ type: 'alert', value: null, alertType: 'duplicateThoughts' }), 2000)
> > ```
> > 
> > 
> > Maybe my solution is overcomplicated, and we just can prevent execution of these lines, if duplicate alert has been already dispatched. In this case after 2 second alert will be dismissed. But If user hits `Enter` again after 1.9s of previous hitting, alert will be dismissed in .1s
> 
> The `alert` action can be dispatched again without consequence. It's the `setTimeout` that's the problem. The old `setTimeout` needs to be cleared before it is set again.

Implemented in b47ffe2

raineorshine commented:
> Implemented in [b47ffe2](https://github.com/cybersemics/em/commit/b47ffe2bc5fb925c77bb40eea5e0a14c7a59d8df)

My apologies, I misinterpreted your previous message and didn't know it was already complete!

raineorshine commented:
> [merge on delete](https://github.com/cybersemics/em/pull/786#issuecomment-662750292)

Could you implement the same alert timing solution for this case as well? Thanks!

Then it should be good to merge. We can do unit tests in a separate PR.

pushkov-fedor commented:
> > [merge on delete](https://github.com/cybersemics/em/pull/786#issuecomment-662750292)
> 
> Could you implement the same alert timing solution for this case as well? Thanks!
> 
> Then it should be good to merge. We can do unit tests in a separate PR.

Done. I decided that adding such alert logic to `alert action-creator` is better approach instead of creating duplicate functions in different parts of project. What do you think?

raineorshine commented:
> Done. I decided that adding such alert logic to `alert action-creator` is better approach instead of creating duplicate functions in different parts of project. What do you think?

Yes, great idea! That's a much better architecture.

raineorshine commented on src/action-creators/alert.ts:
```diff
@@ -4,26 +4,51 @@ import { ActionCreator } from '../types'
+const alert = (value: string | FunctionComponent | null, { alertType, showCloseLink, clearTimeout }: Options = {}): ActionCreator => (dispatch, getState) => {
   const { alert } = getState()
-  if (alert && alert.value === value) return
+  if (alert && alert.value === value) {
+    if (clearTimeout && clearAlertTimeoutId) {
+      window.clearTimeout(clearAlertTimeoutId)
+      clearAlertTimeoutId = window.setTimeout(() => dispatch({
+        type: 'alert',
+        alertType,
+        showCloseLink,
+        value: null,
+      }), clearTimeout)
+    }
+    return
+  }
 
   dispatch({
     type: 'alert',
     alertType,
     showCloseLink,
     value,
   })
+  if (clearTimeout) {
+    clearAlertTimeoutId = window.setTimeout(() => {
+      dispatch({
+        type: 'alert',
+        alertType,
+        showCloseLink,
+        value: null,
+      })
+      clearAlertTimeoutId = null
+    }, clearTimeout)
+  }
```
You've duplicated the code to close the alert. Whenever you are duplicating code that should be a sign that it is not structured properly. Please refactor this to remove the duplication.

Hint: The solution does not involve creating a separate function that gets called in two places. Look at the logic of the action-creator as a whole and you should see that it can be simplified.

raineorshine commented on src/action-creators/alert.ts:
```diff
@@ -4,18 +4,36 @@ import { ActionCreator } from '../types'
+const alert = (value: string | FunctionComponent | null, { alertType, showCloseLink, clearTimeout }: Options = {}): ActionCreator => (dispatch, getState) => {
   const { alert } = getState()
+
+  if (clearTimeout) {
+    // if clearAlertTimeoutId !== null, it means that previous alert hasn't been cleared yet. In this case cancel previous timeout and start new.
+    clearAlertTimeoutId && window.clearTimeout(clearAlertTimeoutId)
+    clearAlertTimeoutId = window.setTimeout(() => {
+      dispatch({
+        type: 'alert',
+        alertType,
+        showCloseLink,
+        value: null,
+      })
+      clearAlertTimeoutId = null
+    }, clearTimeout)
+  }
```
Great!
